src/open_options.rs

```diff
@@ -0,0 +1,139 @@
+use rustc::lint::*;
+use rustc_front::hir::{Expr, ExprMethodCall, ExprLit};
+use utils::{walk_ptrs_ty_depth, match_type, span_lint, OPEN_OPTIONS_PATH};
+use syntax::codemap::{Span, Spanned};
+use syntax::ast::Lit_::LitBool;
+
+declare_lint! {
+    pub NONSENSICAL_OPEN_OPTIONS,
+    Warn,
+    "The options used for opening a file are nonsensical"
+}
+
+
+#[derive(Copy,Clone)]
+pub struct NonSensicalOpenOptions;
+
+impl LintPass for NonSensicalOpenOptions {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(NONSENSICAL_OPEN_OPTIONS)
+    }
+}
+
+impl LateLintPass for NonSensicalOpenOptions {
+    fn check_expr(&mut self, cx: &LateContext, e: &Expr) {
+        if let ExprMethodCall(ref name, _, ref arguments) = e.node {
+            let (obj_ty, _) = walk_ptrs_ty_depth(cx.tcx.expr_ty(&arguments[0]));
+            if name.node.as_str() == "open" && match_type(cx, obj_ty, &OPEN_OPTIONS_PATH){
+                let mut options = Vec::new();
+                get_open_options(cx, &arguments[0], &mut options);
+                check_open_options(cx, &options, e.span);
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+enum Argument {
+    True,
+    False,
+    Unknown
+}
+
+#[derive(Debug)]
+enum OpenOption {
+    Write(Argument),
+    Read(Argument),
+    Truncate(Argument),
+    Create(Argument),
+    Append(Argument)
+}
+
+fn get_open_options(cx: &LateContext, argument: &Expr, options: &mut Vec<OpenOption>) {
+    if let ExprMethodCall(ref name, _, ref arguments) = argument.node {
+        let (obj_ty, _) = walk_ptrs_ty_depth(cx.tcx.expr_ty(&arguments[0]));
+        
+        // Only proceed if this is a call on some object of type std::fs::OpenOptions
+        if match_type(cx, obj_ty, &OPEN_OPTIONS_PATH) && arguments.len() >= 2 {
+            
+            let argument_option = match arguments[1].node {
+                ExprLit(ref span) => {
+                    if let Spanned {node: LitBool(lit), span: _} = **span {
+                        if lit {Argument::True} else {Argument::False}
+                    } else {
+                        return; // The function is called with a literal
```

Comment:
If our argument is truly a std::fs::OpenOptions, this should be practically impossible. Why not either use `unreachable!()` or `Argument::Unknown` if you are unsure?